File: Year2023/CSharp/Day06.cs
using NUnit.Framework;
using System;
using System.Diagnostics;

namespace AdventOfCode.Year2023
{
    [TestFixture]
    [NonParallelizable]
    public class Day06
    {
        private readonly struct Race
        {
            public long TimeInMs { get; }
            public long BestDistance { get; }

            public Race(long time, long distance)
            {
                TimeInMs = time;
                BestDistance = distance;
            }
        }

        [TestCase]
        public void Part01()
        {
            var Input = new Race[]
            {
                new(56, 499),
                new(97, 2210),
                new(77, 1097),
                new(93, 1440)
            };

            int FinalAnswer = 1;

            foreach (Race CurrentRace in Input)
            {
                int WinningCombos = 0;

                for (int IndexMiliseconds = 1; IndexMiliseconds < CurrentRace.TimeInMs; IndexMiliseconds++)
                {
                    // (7 - 2) = (5 * 2) = (10 > 9) == True
                    long DistanceDone = (CurrentRace.TimeInMs - IndexMiliseconds) * IndexMiliseconds;
                    if (DistanceDone > CurrentRace.BestDistance)
                    {
                        WinningCombos++;
                    }
                }
                
                FinalAnswer *= WinningCombos;
            }

            Console.WriteLine("Part one answer: " + FinalAnswer);
            Trace.Assert(FinalAnswer == 1710720, $"Expected final answer to be 1710720 but it was {FinalAnswer}");
        }

        [TestCase]
        public void Part02()
        {
            long FinalAnswer;
            int WinningCombos = 0;
            Race CurrentRace = new(56977793, 499221010971440);

            // Seconds to hold * Max time - Seconds to hold squared > distance
            // S*T - S^2 > D
            for (int IndexMiliseconds = 1; IndexMiliseconds < CurrentRace.TimeInMs; IndexMiliseconds++)
            {
                long DistanceDone = (CurrentRace.TimeInMs - IndexMiliseconds) * IndexMiliseconds;
                if (DistanceDone > CurrentRace.BestDistance)
                {
                    WinningCombos++;
                }
            }

            FinalAnswer = WinningCombos;

            Console.WriteLine("Part two answer: " + FinalAnswer);
            Trace.Assert(FinalAnswer == 35349468, $"Expected final answer to be 35349468 but it was {FinalAnswer}");
        }
        
        [TestCase]
        public void Part02_Variation01()
        {
            long FinalAnswer;
            int WinningCombos = 0;
            Race CurrentRace = new(56977793, 499221010971440);

            // I know this thing is a quadratic equation but I massively suck at math so I always try for logical optimizations instead
            // The pattern I noticed is that the winning times appear near the middle so I reckon there's a curve of some kind.
            // Hence this optimization is to split the time in two, then with two loops go down and then until there are no winning moves anymore.
            var TimeInMS = CurrentRace.TimeInMs / 2;

            for (long IndexMiliseconds = TimeInMS; IndexMiliseconds > 1; IndexMiliseconds--)
            {
                long DistanceDone = (CurrentRace.TimeInMs - IndexMiliseconds) * IndexMiliseconds;
                if (DistanceDone > CurrentRace.BestDistance)
                {
                    WinningCombos++;
                    continue;
                }

                break;
            }

            for (long IndexMiliseconds = TimeInMS+1; IndexMiliseconds < CurrentRace.TimeInMs; IndexMiliseconds++)
            {
                long DistanceDone = (CurrentRace.TimeInMs - IndexMiliseconds) * IndexMiliseconds;
                if (DistanceDone > CurrentRace.BestDistance)
                {
                    WinningCombos++;
                    continue;
                }

                break;
            }

            FinalAnswer = WinningCombos;

            Console.WriteLine("Part two answer: " + FinalAnswer);
            Trace.Assert(FinalAnswer == 35349468, $"Expected final answer to be 35349468 but it was {FinalAnswer}");
        }

    }
}

File: Year2023/C/src/Day06.c
#include "LibThomas.h"
#include "stdio.h"
#include <assert.h>
#include <stdint.h>
#include <strings.h>

typedef struct _Race {
    int64_t TimeInMS;
    int64_t BestDistance;
} Race;

int main(void)
{
    Race Part01Input[4] = {
        { .TimeInMS = 56, .BestDistance = 499 },
        { .TimeInMS = 97, .BestDistance = 2210 },
        { .TimeInMS = 77, .BestDistance = 1097 },
        { .TimeInMS = 93, .BestDistance = 1440 }
    };

    int32_t FinalAnswerPart1 = 1;
    int32_t FinalAnswerPart2 = 0;

    for (size_t Index = 0; Index < arrayCount(Part01Input); Index++) {

        int32_t WinningCombos = 0;
        Race CurrentRace = Part01Input[Index];

        for (int32_t IndexMiliseconds = 1; IndexMiliseconds < CurrentRace.TimeInMS; IndexMiliseconds++)
        {
            int64_t DistanceDone = (CurrentRace.TimeInMS - IndexMiliseconds) * IndexMiliseconds;
            if (DistanceDone > CurrentRace.BestDistance)
            {
                WinningCombos++;
            }
        }

        FinalAnswerPart1 *= WinningCombos;
    }

    printf("Part 1 answer: %i\n", FinalAnswerPart1);
    assert(FinalAnswerPart1 == 1710720);

    // Part 02

    Race Part02Input = { .TimeInMS = 56977793, .BestDistance = 499221010971440 };
    int32_t WinningCombos = 0;
    int64_t TimeInMS = Part02Input.TimeInMS / 2;

    for (int64_t IndexMiliseconds = TimeInMS; IndexMiliseconds > 1; IndexMiliseconds--)
    {
        int64_t DistanceDone = (Part02Input.TimeInMS - IndexMiliseconds) * IndexMiliseconds;
        if (DistanceDone > Part02Input.BestDistance)
        {
            WinningCombos++;
            continue;
        }

        break;
    }

    for (int64_t IndexMiliseconds = TimeInMS+1; IndexMiliseconds < Part02Input.TimeInMS; IndexMiliseconds++)
    {
        int64_t DistanceDone = (Part02Input.TimeInMS - IndexMiliseconds) * IndexMiliseconds;
        if (DistanceDone > Part02Input.BestDistance)
        {
            WinningCombos++;
            continue;
        }

        break;
    }

    FinalAnswerPart2 = WinningCombos;
    printf("Part 2 answer: %i\n", FinalAnswerPart2);
    assert(FinalAnswerPart2 == 35349468);
}
